cognite/client/api/assets.py

```diff
@@ -411,3 +454,146 @@ def search(
+        ref_ids_set = set(ref_ids)
+        assert sorted(ref_ids) == sorted(list(ref_ids_set)), "Duplicate ref_ids found"
+        for asset in assets:
+            parent_ref = asset.parent_ref_id
+            if parent_ref:
+                assert parent_ref in ref_ids_set, "parent_ref_id '{}' does not point to anything".format(parent_ref)
+                assert asset.parent_id is None, "An asset has both parent_id and parent_ref_id set."
+
+    @staticmethod
+    def initialize_ref_id_to_remaining_children_map(assets: List[Asset]) -> Dict[str, List[Asset]]:
+        ref_id_to_children = {asset.ref_id: [] for asset in assets if asset.ref_id is not None}
+        for asset in assets:
+            if asset.parent_ref_id in ref_id_to_children:
+                ref_id_to_children[asset.parent_ref_id].append(asset)
+        return ref_id_to_children
+
+    def get_descendants_unblocked_by_ref_id(self, assets: List[Asset], limit):
+        unblocked_children = []
+        for asset in assets:
+            if asset.ref_id in self.ref_id_to_remaining_children:
+                asset_children = self.ref_id_to_remaining_children[asset.ref_id].copy()
+                for child_asset in asset_children:
+                    unblocked_children.append(child_asset)
+                    self.ref_id_to_remaining_children[asset.ref_id].remove(child_asset)
+                    if len(unblocked_children) == limit:
+                        return unblocked_children
+
+                    max_descendents = limit - len(unblocked_children)
+                    unblocked_children.extend(self.get_descendants_unblocked_by_ref_id([child_asset], max_descendents))
+                    if len(unblocked_children) == limit:
+                        return unblocked_children
+        return unblocked_children
+
+    def get_unblocked_assets(self, limit) -> List[List[Asset]]:
+        unblocked_assets_lists = []
+        unblocked_assets = []
+
+        for asset in self.remaining_assets:
+            if asset.parent_ref_id is None or asset.parent_id is not None:
+                unblocked_assets.append(asset)
+            elif self.ref_id_to_id.get(asset.parent_ref_id) is not None:
+                asset.parent_id = self.ref_id_to_id[asset.parent_ref_id]
+                unblocked_assets.append(asset)
+            if len(unblocked_assets) == limit:
+                unblocked_assets_lists.append(unblocked_assets)
+                unblocked_assets = []
+
+        if len(unblocked_assets) < limit:
+            max_descendents = limit - len(unblocked_assets)
+            unblocked_assets.extend(self.get_descendants_unblocked_by_ref_id(unblocked_assets, max_descendents))
+
+        if len(unblocked_assets) > 0:
+            unblocked_assets_lists.append(unblocked_assets)
+
+        for unblocked_assets in unblocked_assets_lists:
+            for unblocked_asset in unblocked_assets:
+                self.remaining_assets.remove(unblocked_asset)
```

Comment:
This will probably become too expensive when you have many assets (e.g. doing linear search and linear shift over a million items several thousand times). Maybe make `self.remaining_assets` a set?